File: day:59/maxPathSum.py
import sys

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, root, maxSum):
        if(root is None):
            return 0;
    
        left = self.dfs(root.left, maxSum);
        right = self.dfs(root.right, maxSum);
    
        maxSum[0] = max(maxSum[0], root.val + left + right);
        return max(0, root.val + max(left, right));
    
    def maxPathSum(self, root: TreeNode) -> int:
        maxSum = [-sys.maxsize - 1];
        self.dfs(root, maxSum);
        
        return maxSum[0];
        

File: day:59/maxPathSum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */



function dfs(root, maxSum) {
    if(root === null) {
        return 0;
    }
    
    let left = dfs(root.left, maxSum);
    let right = dfs(root.right, maxSum);
    
    maxSum[0] = Math.max(maxSum[0], root.val + left + right);
    return Math.max(0, root.val + Math.max(left, right));
}

/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function(root) {
    let maxSum = [-Infinity];
    dfs(root, maxSum);
    return maxSum[0];
    
};
